src/main/scala/com/microsoft/ml/spark/lightgbm/booster/LightGBMBooster.scala

```diff
@@ -167,35 +182,155 @@ protected class BoosterHandler(model: String) {
     out
   }
 
-  private def freeNativeMemory(): Unit = {
-    if (boosterPtr != null) {
-      LightGBMUtils.validate(lightgbmlib.LGBM_BoosterFree(boosterPtr), "Finalize Booster")
-      boosterPtr = null
-    }
-  }
-
   override protected def finalize(): Unit = {
     freeNativeMemory()
     super.finalize()
   }
 }
 
 /** Represents a LightGBM Booster learner
-  * @param model The string serialized representation of the learner
+  * @param trainDataset The training dataset
+  * @param parameters The booster initialization parameters
+  * @param modelStr Optional parameter with the string serialized representation of the learner
   */
 @SerialVersionUID(777L)
-class LightGBMBooster(val model: String) extends Serializable {
-  /** Transient variable containing local machine's pointer to native booster
+class LightGBMBooster(val trainDataset: Option[LightGBMDataset] = None, val parameters: Option[String] = None,
+                      modelStr: Option[String] = None) extends Serializable {
+
+  /** Represents a LightGBM Booster learner
+    * @param trainDataset The training dataset
+    * @param parameters The booster initialization parameters
     */
+  def this(trainDataset: LightGBMDataset, parameters: String) = {
+    this(Some(trainDataset), Some(parameters))
+  }
+
+  /** Represents a LightGBM Booster learner
+    * @param model The string serialized representation of the learner
+    */
+  def this(model: String) = {
+    this(modelStr = Some(model))
+  }
+
   @transient
   lazy val boosterHandler: BoosterHandler = {
-    new BoosterHandler(model)
+    val boosterOutPtr = lightgbmlib.voidpp_handle()
+    if (trainDataset.isEmpty && model.isEmpty) {
+      throw new IllegalArgumentException("One of training dataset or serialized model parameters must be specified")
+    }
+    if (trainDataset.isEmpty) {
+      new BoosterHandler(model)
+    } else {
+      LightGBMUtils.validate(lightgbmlib.LGBM_BoosterCreate(trainDataset.map(_.datasetPtr).get,
+        parameters.get, boosterOutPtr), "Booster")
+      new BoosterHandler(lightgbmlib.voidpp_value(boosterOutPtr))
+    }
   }
 
+  val model = modelStr.getOrElse("")
```

Comment:
what is the behavior if this is ""? Seems this should be kept as option especially if code elsewhere checks for ""

Reply:
model string should never be empty

Reply:
removed the model